src/libNode/Node.cpp

```diff
@@ -246,265 +286,33 @@ bool Node::CheckState(Action action)
-                      "Unrecognized or error state");
-            result = false;
-            break;
-        }
-        break;
-    case PROCESS_FINALBLOCK:
-        switch (m_state)
-        {
-        case POW1_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "POW1_SUBMISSION");
-            result = false;
-            break;
-        case POW2_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "POW2_SUBMISSION");
-            result = false;
-            break;
-        case TX_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in TX_SUBMISSION");
-            result = false;
-            break;
-        case TX_SUBMISSION_BUFFER:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "TX_SUBMISSION_BUFFER");
-            result = false;
-            break;
-        case MICROBLOCK_CONSENSUS_PREP:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "MICROBLOCK_CONSENSUS_PREP");
-            result = false;
-            break;
-        case MICROBLOCK_CONSENSUS:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "MICROBLOCK_CONSENSUS");
-            result = false;
-            break;
-        case WAITING_FINALBLOCK:
-            break;
-        case ERROR:
-            LOG_GENERAL(WARNING,
-                        "Doing WAITING_FINALBLOCK but receiving ERROR message");
-            result = false;
-            break;
-        default:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Unrecognized or error state");
-            result = false;
-            break;
-        }
-        break;
-    default:
         LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
                   "Unrecognized action");
-        result = false;
-        break;
+        return false;
     }
 
-    return result;
+    std::ostringstream error;
+    if (m_state == ERROR)
+    {
+        error << "Doing " << ActionString(action)
+              << " but receiving ERROR message";
+        LOG_GENERAL(WARNING, error.str());
+        return false;
+    }
+
+    error << "Doing " << ActionString(action) << " but already in "
+          << NodeStateString(m_state);
+    LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
+              error.str());
+
```

Comment:
LOG_XXX already support operator<< inside.
It's needless to use another ostringstream to do it and that is, we can use following:
LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(), "Doing " << ActionString(action) << " but already in " << NodeStateString(m_state));

Reply:
Great tip, thanks!